### DISCUSSION TITLE
[needs help] Celery & Selenium — errors when calling selenium in celery task

### DISCUSSION THREAD

**Discussion by ekalinin** [celery/celery#7445] [Q&A]
Hi,

I have strange situation with calling selenium within celery task.
Source code:

```
from pyvirtualdisplay import Display
from selenium import webdriver
from celery.task import task

@task
def get():

    display = Display(visible=0, size=(800, 600))
    display.start()

    b = webdriver.Firefox()
    return 1

@task
def get_stub():

    display = Display(visible=0, size=(800, 600))
    display.start()

    return 1
```

When i call task in sync mode there is no errors and it's all good
(executing from my os user):

```
python -c "import mymodule.simple; print mymodule.simple.get()";
1
```

But when i tried to call selenium in async celery task i am getting an error
(executing from 'celery' user):

```
python -c "import mymodule.simple; print mymodule.simple.get.delay().wait()";
...
WebDriverException: Message: None 
```

Traceback from celery.log:

```
  File "/usr/local/lib/python2.6/dist-packages/celery/execute/trace.py", line 47, in trace
    return cls(states.SUCCESS, retval=fun(*args, **kwargs))
  File "/usr/local/lib/python2.6/dist-packages/celery/app/task/__init__.py", line 247, in __call__
    return self.run(*args, **kwargs)
  File "/usr/local/lib/python2.6/dist-packages/celery/app/__init__.py", line 175, in run
    return fun(*args, **kwargs)
  File "/.../simple.py", line 11, in get
    b = webdriver.Firefox()
  File "/usr/local/lib/python2.6/dist-packages/selenium/webdriver/firefox/webdriver.py", line 45, in __init__
    self.binary, timeout),
  File "/usr/local/lib/python2.6/dist-packages/selenium/webdriver/firefox/extension_connection.py", line 46, in __init__
    self.binary.launch_browser(self.profile)
  File "/usr/local/lib/python2.6/dist-packages/selenium/webdriver/firefox/firefox_binary.py", line 44, in launch_browser
    self._wait_until_connectable()
  File "/usr/local/lib/python2.6/dist-packages/selenium/webdriver/firefox/firefox_binary.py", line 81, in _wait_until_connectable
    self._get_firefox_output())
WebDriverException: Message: None 
```

Celery works fine:

```
python -c "import mymodule.simple; print mymodule.get_stub.delay().wait()";
1
```

Installed soft:
- celery 2.4.6
- headless firefox
- selenium (selenium-server-standalone-2.11.0.jar)

What am I doing wrong? What prevents selenium work? Any ideas?

Thanks.

**Comment by ask** [comment]
Are you running celeryd daemonized?  In that case it could be due to not being attached to a tty, or it could be because
the python selenium library is initialized in the parent process, then that is carried over to the child process causing problems.

You can try two things,

1) Run celeryd from a terminate (don't daemonize)
2) If that doesn't work, try running celeryd with -P solo

If using -P solo works, then the problem is probably due to initialization being done too early,
and you should try delaying importing anything from selenium until the task actually runs:

```
 @task
 def get():
      from selenium import webdriver
```

(Note that anything that imports selenium before until the first task actually runs may
affect this)

**Comment by ekalinin** [comment]
Hi, 

Yes i used celeryd daemonized.
Just now tried to run celeryd from the terminal — success! No errors.

Also, just checked that in my case fit this script: `/contrib/debian/init.d-deprecated/celeryd` (instead of `/contrib/generic-init.d/celeryd`).

Thanks a lot!

**Comment by palewire** [comment]
I'm having this same issue and I'm not 100% clear on what the fix is. Can someone dumb it down for me?

**Comment by ask** [comment]
I'm guessing the Python selenium lib is not fork-safe.

For example it could be that when you import selenium in the parent process it initializes some global variables
that no longer works when the child process "inherits" the parents memory.

The "moving the imports to runtime" solution is a hack, and it should really be considered a bug in Selenium, and they should fix it.

**Comment by palewire** [comment]
Thank you. I will try that. 

Just to give you some more information. This kind of task worked great for me in development on my laptop, when I ran celeryd from the manage.py command in a separate terminal. But when I tried to deploy it to a headless production machine and deamonized celery is when I ran into the problems.

**Comment by palewire** [comment]
Okay, so I moved the selenium import inside my task.

``` python
@task()
def get_screenshot(site_id, update_id):
    """
    Create a screenshot and save it to the database
    """
    from selenium import webdriver
    # Moar...
```

 I have the approved celeryd init.d script, with my Django settings at the top.

``` bash
CELERYD_CHDIR="/apps/pastpages.org/repo"
ENV_PYTHON="/apps/pastpages.org/bin/python"
CELERYD_MULTI="$ENV_PYTHON $CELERYD_CHDIR/manage.py celeryd_multi"
CELERYCTL="$ENV_PYTHON $CELERYD_CHDIR/manage.py celeryctl"
CELERYD_OPTS="--purge"
CELERY_CONFIG_MODULE="celeryconfig"
export DJANGO_SETTINGS_MODULE="project.settings"
```

And after I run `celeryd start` I fire off a task that queues up 10 `get_screenshot` tasks. The initial queue job works fine, but the screenshots start to fail with the following.

`````` bash
Traceback (most recent call last):
  File "/apps/pastpages.org/lib/python2.7/site-packages/celery/execute/trace.py", line 34, in trace
    return cls(states.SUCCESS, retval=fun(*args, **kwargs))
  File "/apps/pastpages.org/lib/python2.7/site-packages/celery/task/base.py", line 241, in __call__
    return self.run(*args, **kwargs)
  File "/apps/pastpages.org/lib/python2.7/site-packages/celery/app/__init__.py", line 141, in run
    return fun(*args, **kwargs)
  File "/apps/pastpages.org/repo/archive/tasks.py", line 67, in get_screenshot
    encoded_data = browser.get_screenshot_as_base64()
  File "/apps/pastpages.org/lib/python2.7/site-packages/selenium/webdriver/remote/webdriver.py", line 697, in get_screenshot_as_base64
    return self.execute(Command.SCREENSHOT)['value']
  File "/apps/pastpages.org/lib/python2.7/site-packages/selenium/webdriver/remote/webdriver.py", line 153, in execute
    self.error_handler.check_response(response)
  File "/apps/pastpages.org/lib/python2.7/site-packages/selenium/webdriver/remote/errorhandler.py", line 147, in check_response
    raise exception_class(message, screen, stacktrace)
WebDriverException: Message: None 
None```
``````

**Comment by palewire** [comment]
Just in the interest of documenting all this for future generations, here's what the [python-selenium FAQ](http://code.google.com/p/selenium/wiki/FrequentlyAskedQuestions#Q:_Is_WebDriver_thread-safe?) says on the matter:

<blockquote>Q: Is WebDriver thread-safe?
A: WebDriver is not thread-safe. Having said that, if you can serialise access to the underlying driver instance, you can share a reference in more than one thread. This is not advisable. You /can/ on the other hand instantiate one WebDriver instance for each thread.</blockquote>

**Comment by sadjad** [comment]
I'm exactly having the same problem. I use **supervisord** to run Celery as a daemon and Selenium won't work in that case. But when I run Celery directly from terminal (not as a daemon), it works okay. Do you have any ideas how I can solve this problem and still use supervisord?

**Comment by adamtimmins** [comment]
Has there been any fix to this to get Celery working with supervisord and Selenium? 

Importing the selenium into the task has no effect. 

I'm running the task successfully on the command line without issue but I'd like to use the recommended supervisord config.

**Comment by BoPeng** [comment]
I would also appreciate an answer. Similar to others, the application is to use selenium to take screenshots of URLs passed to a celery task. The task works ok in tty, but fails to take any screenshot when selenium is started from celery tasks started by supervisord.

**Comment by ashmitsharma** [comment]
Can some one please tell me the solution to this problem I need it ASAP. Please help.
